[export][dynamic shapes] add Dim._OBLIVIOUS, _mark_oblivious()

pianpwk opened the pull request:
Summary: Adds `Dim._OBLIVIOUS` in export dynamic shapes, and `_mark_oblivious()` in dynamo decorators, to support the use of OBLIVIOUS_SIZE.

The semantics are that we allocate what looks like a unbacked symbol, but is technically backed; it contains a hint, the user-intention is just to opt into size-oblivious reasoning and avoid 0/1 specialization.

Decided to do this over mark_unbacked + hint because it's easier to write code in symbolic_shapes that distinguishes between valid reasoning for oblivious sizes and general unbacked (e.g. we can set replacements for oblivious sizes since they're graph inputs, and we don't face the "time traveling" problem): https://github.com/pytorch/pytorch/blob/3a69dee955f2c6c57f7c879ba82469fa0c1d0b74/torch/fx/experimental/symbolic_shapes.py#L6284-L6297

On the other hand if we just handled this with unbacked + hint, it's hard to tell if we're dealing with input sizes that should have hints, or if we've just been doing real-tensor prop.

Test Plan: test_export

Differential Revision: D70193972




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

angelayi commented on torch/export/dynamic_shapes.py:
```diff
@@ -46,11 +46,13 @@ class _DimHint(Enum):
     - AUTO means automatic inference of shape (static or dynamic).
     - STATIC means static shape (always specialized).
     - DYNAMIC means dynamic, will error out if specialized.
+    - UNBACKED means allocating an unbacked symbol (with hint).
```
Can you clarify the behavior of Dim.UNBACKED? In general I feel this will just add more confusion to users who want to specify dynamism.

My understanding is that this is like DYNAMIC but w/o 0/1 specialization, however below you did `mark_unbacked(..., strict=False)` meaning that it won't error if this is specialized. So that means the behavior is more like AUTO. But at the same time why can't AUTO just use mark_unbacked.

pianpwk commented on test/dynamo/test_misc.py:
```diff
@@ -10877,6 +10877,20 @@ def fn(x, y):
+        f(x, y)
+        f(torch.randn(2), torch.randn(1))
+        f(torch.randn(0), torch.randn(0))
+
```
just testing 0/1 specialization

pianpwk commented on test/export/test_export.py:
```diff
@@ -2351,6 +2351,78 @@ def forward(self, x, y, z):
+        ph_x, ph_y = [node for node in ep.graph.nodes][:2]
+        u0 = ph_x.meta["val"].shape[0].node.expr
+        u1 = ph_y.meta["val"].shape[0].node.expr
+        self.assertEqual(shape_env.replacements.get(u1, u1), u0)
```
test we allow replacements on oblivious sizes

pianpwk commented on test/export/test_export.py:
```diff
@@ -2351,6 +2351,78 @@ def forward(self, x, y, z):
+        ep = export(Foo(), inps, dynamic_shapes=shapes)
+        ep.module()(torch.randn(1, 1, 0))
+        ep.module()(torch.randn(5, 5, 5))
+        ep.module()(torch.randn(0, 0, 1))
```
tests exporting with 0/1 sizes

pianpwk committed:
[export][dynamic shapes] add Dim._OBLIVIOUS, _mark_oblivious() (#147881)

Summary:
Adds `Dim._OBLIVIOUS` in export dynamic shapes, and `_mark_oblivious()` in dynamo decorators, to support the use of OBLIVIOUS_SIZE.

The semantics are that we allocate what looks like a unbacked symbol, but is technically backed; it contains a hint, the user-intention is just to opt into size-oblivious reasoning and avoid 0/1 specialization.

Decided to do this over mark_unbacked + hint because it's easier to write code in symbolic_shapes that distinguishes between valid reasoning for oblivious sizes and general unbacked (e.g. we can set replacements for oblivious sizes since they're graph inputs, and we don't face the "time traveling" problem: https://github.com/pytorch/pytorch/blob/3a69dee955f2c6c57f7c879ba82469fa0c1d0b74/torch/fx/experimental/symbolic_shapes.py#L6286).

On the other hand if we just handled this with unbacked + hint, it's hard to tell if we're dealing with input sizes that should have hints, or if we've just been doing real-tensor prop.


Test Plan:
test_export





cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: D70193972

Pulled By: pianpwk

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -846,26 +846,45 @@ def has_free_symbols(val: IterateExprs) -> bool:
+            and not (
+                exclude_oblivious and arg in shape_env.oblivious_var_to_val
+            )
+        )
```
these symbols appear unbacked, but they're technically backed, so I guess what you want from this function is contextual?

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6318,7 +6343,7 @@ def trivial_solve(lhs: sympy.Expr, rhs: sympy.Expr) -> bool:
                     ):
                         new_var = self._find(r[1])
-                        ok = len(free_unbacked_symbols(new_var)) == 0
+                        ok = not has_free_unbacked_symbols(new_var, exclude_oblivious=True, shape_env=self)
```
we won't face the time-traveling problem mentioned in the PR description because these are all input sizes to the graph, so it should be ok to set replacements

bobrenjc93 commented on torch/_dynamo/decorators.py:
```diff
@@ -511,6 +511,26 @@ def mark_unbacked(t, index, strict=False):
+def _mark_oblivious(t, index):
+    """
+    Mark a tensor as having a size-oblivious dim. This follows unbacked semantics for 0/1 specialization,
+    but allows guards on the symbol (the symbol has a hint, so is technically backed).
```
nit "so it is technically backed"

bobrenjc93 commented on torch/export/dynamic_shapes.py:
```diff
@@ -46,11 +46,13 @@ class _DimHint(Enum):
     - AUTO means automatic inference of shape (static or dynamic).
     - STATIC means static shape (always specialized).
     - DYNAMIC means dynamic, will error out if specialized.
+    - _OBLIVIOUS allocates an backed symbol with size-oblivious semantics.
```
"a backed" not "an backed"

also is the _ prefix intentional?

pianpwk replied:
yeah, making this ugly and private for now lol, not committing to it

bobrenjc93 commented:
I'm still reviewing, but my general instinct is in line with @angelayi. 

I was expecting to see something like Dim.UNBACKED where under the hood we call

torch.dynamo.decorators.mark_unbacked()
torch.dynamo.decorators.check_is_size()

Can you spell out an explicit example as to why this doesn't work?

bobrenjc93 commented:
In the context of export, there really is no difference between a runtime assert and a guard right?

It feels like what you really want is an unbacked symint that you can add runtime asserts on? Basically right now instead of throwing guard on data dependent, if we inserted a runtime assert then and there, it's effectively what you want?

I'd much prefer we have an assert_on_data_dependent mode/flag/etc instead of introducing more obliviousness, especially given the already high amount of confusion with size-obliviousness and general momentum towards more explicit semantics.

I'll sleep on it, but this might be worth doing some collective design/whiteboarding. These API decisions are hard to revert, so I want to make sure we do them right.

ColinPeppler commented:
Here's some motivation for this PR. as an Export user, I'd want to Export that this dynamic dim can be sizes 0/1 -- or the bounds for this guy is `[0, inf]`.

**Why?** Because I may want to use the lowerbound in Inductor/AOTI for size assertions (disabled by default in AOTI) or for runtime assertions on unbacked symints.

xref: https://fb.workplace.com/groups/1028545332188949/permalink/1130301062013375/

pianpwk commented on test/export/test_export.py:
```diff
@@ -2351,6 +2351,78 @@ def forward(self, x, y, z):
+
+        class Foo(torch.nn.Module):
+            def forward(self, x, y):
+                assert x.shape[0] == y.shape[0]
```
this was meant to be an assert, adding a guard we need to resolve

pianpwk commented:
for @bobrenjc93's points

> I was expecting to see something like Dim.UNBACKED where under the hood we call
>
> torch.dynamo.decorators.mark_unbacked()
> torch.dynamo.decorators.check_is_size()
>
> Can you spell out an explicit example as to why this doesn't work?

What we really need in export is the backing hint, so that if we hit guards, we can specialize on the guard to produce a graph (since we can't graph break). Without the hint this introduces too many data-dependent errors, and the benefit of avoiding 0/1 specialization is barely worth the cost (torch._checks everywhere), when those checks weren't needed with a normal backed symbol.

The initial implementation (Dim.UNBACKED) essentially did this:
```
torch.dynamo.decorators.mark_unbacked()
torch.dynamo.decorators.check_is_size()
shape_env.set_unbacked_var_to_val(s, val)
```

At a fundamental level, this works, and isn't so different (this PR just does mark_oblivious() and stores the value in oblivious_var_to_val). I ended up doing this because:
- If you read the original issue description for [OBLIVIOUS_SIZE](https://github.com/pytorch/pytorch/issues/137100), it's very much inline with what @ColinPeppler wants
- unbacked_var_to_val works for basic purposes, but differs from backed symbols in what's valid reasoning in symbolic_shapes, and it's easier to write that reasoning using oblivious_var_to_val, knowing the assumptions that oblivious_var_to_val makes, versus using a lot of special casing in unbacked_var_to_val.

For example with the time-traveling problem: we hit a data-dependent error with the original implementation, because we can't set replacements with unbacked symints, because the free symbols of the replacement expression might not be created in the graph yet, when the replaced symbol gets generated (we're time traveling forward). But the unbacked symints we're interested in replacing are input symbols that we marked with Dim.UNBACKED, and because they're input symbols, they're all present in the graph at the start, so it's ok to set replacements on them.

In this case it ends up being easier, and arguably more coherent in symbolic_shapes code, to say, we'll allow replacements on OBLIVIOUS_SIZE symbols, which have the invariant that they're input size symbols, instead of banning replacements for all unbacked EXCEPT if the target expression free symbols are all input size symbols (which we don't have the info for in create_symbol).

There might be some other minor things I'm forgetting, but this is the main one.

> I'd much prefer we have an assert_on_data_dependent mode/flag/etc instead of introducing more obliviousness, especially given the already high amount of confusion with size-obliviousness and general momentum towards more explicit semantics.

> I'll sleep on it, but this might be worth doing some collective design/whiteboarding. These API decisions are hard to revert, so I want to make sure we do them right.

Yeah, this is why I'm going the ugly, private flag route, to unblock Colin. The API design's not final. I do think whatever we end up offering, there'll be a user need for symbols that don't data-dep error out and don't 0/1 specialize, regardless of if we call them backed or unbacked or if size-oblivious reasoning is still a thing; I mainly see that being a UX change with basically the same underlying mechanics.

cc @laithsakka @avikchaudhuri
